### PR TITLE
Introduce VideoPress content gating

### PR DESCRIPTION
**Pull request by pgk**
Fixes https://github.com/Automattic/greenhouse/issues/1808

## Proposed changes:

* Alters the behavior of private videos when embedded within blocks/posts that restrict content to subscribers.
* ~Introduces a way for blocks to control the unauthorized message the video player will display.~
* The video iframe and it's files will not be available if a logged in user has no paid subscription, even if logged in.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion

pxWta-1VK-p2

## Does this pull request change what data or activity we track or use?
no

## Testing instructions:

* You need the Jetpack plugin to test this, as all Earn blocks are Jetpack-plugin-only.
* Do a full rebuild: `pnpm jetpack build plugins/jetpack --deps`
* Refer to pxWta-1VK-p2 on what to sandbox and what not.
* You will need to setup paid subscriptions and paid newsletter subscriptions.
* You will need 1 simple user, 1 newsletter sub user and 1 paid sub user.
* You will need to upgrade to a plan that includes more than one VideoPress videos. (because you are sandboxing store, you might need to do that again)
* On your test site, ensure you can access the paid content block.
* If you haven't, connect to stripe. If you sandboxed store, you can choose to use a "test" connection when on stripe's site.
* You are connected to test stripe now.

### Paid content block

* Go ahead and create a new post and add a "Paid Content" block.
* Assign a subscription plan.
* On the Subscriber view, create a VideoPress block and upload a video. Set privacy to private.
* Save.
* View as admin/editor. You can see the video.
* Try to view as a non-paid subscriber. You should not be able to view.
* View as a paid subscriber. You can view content and video.

### Newsletter

* Enable Newsletters (Jetpack -> Settings -> Newsletters). Create a newsletter subscription.
* Create a new post and set access to paid subscribers only.
* Add a videopress video, set to private. Save and publish.
* View as non-subscriber. You cant access anything.
* View as subscriber. You can see the video.

### Paywall

* Enable Newsletters (Jetpack -> Settings -> Newsletters). Create a newsletter subscription.
* Create post and upload a Video. Set it to public.
* Add a paywall block and upload another video. Set it to private.
* Publish post.
* View as guest and subscriber. You only see the video if a subscriber.

### CODE REVIEW

File: projects/packages/videopress/src/class-ajax.php
Comment (thedebian) [inline projects/packages/videopress/src/class-ajax.php]: Should it somehow reset the postdata afterwards ?
Looking up the documentation, it seems to be something along this line :
`$post = setup_postdata( $post_to_check );`
And later on
`wp_reset_postdata();`

File: projects/packages/videopress/src/class-initializer.php
Comment (thedebian) [inline projects/packages/videopress/src/class-initializer.php]: $guid should probably be esc_js-ed (maybe $premium_block_plan_id also, even though it's casted now, we never know what it'll be in the future :D )

Comment (pgk) [reply projects/packages/videopress/src/class-initializer.php]: Good idea! done

File: projects/packages/videopress/src/class-ajax.php
Comment (pgk) [reply projects/packages/videopress/src/class-ajax.php]: Since this is a wp_ajax call, I think it is OK as the request is specifically about the specific post_id. On the other hand it doesn't hurt to cleanup.

**Comment by roundhill** [comment]
> View as a paid subscriber. You can view content and video.

This is not working for me. I see a token in the query string, but then the console has an error `'Token is not achievable'`

Screenshot 2023-09-13 at 5 56 07 PM

What am I missing? :D

File: projects/packages/videopress/src/class-access-control.php
Comment (roundhill) [inline projects/packages/videopress/src/class-access-control.php]: Yoda-flip this?

File: projects/packages/videopress/src/class-access-control.php
Comment (roundhill) [inline projects/packages/videopress/src/class-access-control.php]: This `return` could be removed since the same call is made outside the if statement with the same vars.

File: projects/packages/videopress/src/class-access-control.php
Comment (roundhill) [inline projects/packages/videopress/src/class-access-control.php]: Same duplicate return statements here.

File: projects/packages/videopress/src/class-access-control.php
Comment (roundhill) [inline projects/packages/videopress/src/class-access-control.php]: Would this be a good place to ensure the video is marked as private, since we know we are dealing with subscriber content here? We could use the `update_video_restriction_details_on_wpcom` function?

**Review by roundhill — CHANGES_REQUESTED** [review]
Overall working well, found a few things!

**Comment by roundhill** [comment]
Dropping notes on how to test this on wp.com:

* Apply D122060-code to your sandbox.
* Run `bin/jetpack-downloader test jetpack add/videopress-content-gating`
* Sandbox your dev site and `public-api`, `subscribe`, `video` subdomains
* Apply additional configurations detailed here to your sandbox: pxWta-1VK#comment-6259-p2

File: projects/packages/videopress/src/class-access-control.php
Comment (pgk) [reply projects/packages/videopress/src/class-access-control.php]: @roundhill do you mean force-setting the privacy setting to private?

Comment (roundhill) [reply projects/packages/videopress/src/class-access-control.php]: Yeah, I was thinking if we make it to this check maybe we could force the video to be set to private so that the JWTs  get requested?

**Review by roundhill — DISMISSED** [review]
Latest changes LGTM!

**Comment by simison** [comment]
Needs a rebase.

File: projects/packages/videopress/src/class-access-control.php
Comment (jeherve) [inline projects/packages/videopress/src/class-access-control.php:86]: This function relies on checking for existing functions. I wonder if it should rely on existing utilities managing editor extension availability, as they're defined in `Jetpack_Gutenberg`, such as `is_available()`?

Should it also take module availability into account, and check if the Subscriptions module is active for example?

File: projects/packages/videopress/src/class-access-control.php
Comment (jeherve) [inline projects/packages/videopress/src/class-access-control.php:227]: I would recommend adding docblocks to the different filters you add in here, so our future selves know how they can be used, and so the [Jetpack Codex](https://developer.jetpack.com/) remains up to date.

File: projects/packages/videopress/src/class-access-control.php
Comment (jeherve) [inline projects/packages/videopress/src/class-access-control.php:101]: Are there situations where it would not exist? I'm not sure I understand what this checks?

File: projects/packages/videopress/src/class-access-control.php
Comment (jeherve) [inline projects/packages/videopress/src/class-access-control.php]: Would it be an option to use `read_post` here instead of relying on editing the `$post` global?

https://github.com/WordPress/wordpress-develop/blob/a55dcf46b8d3ce93affed8fbeb9c6a2645157f3f/src/wp-includes/capabilities.php#L278

**Review by jeherve — COMMENTED** [review]
Just a few questions, but I'll let the folks on team Zap look at this, they know more about Jetpack Memberships than I do!

**Comment by simison** [comment]
> but I'll let the folks on team Zap look at this, they know more about Jetpack Memberships than I do!

That would be @Automattic/gold for memberships and I believe the teams collaborated on this.

File: projects/packages/videopress/src/class-access-control.php
Comment (millerf) [inline projects/packages/videopress/src/class-access-control.php:107]: That would be always false as https://github.com/Automattic/jetpack/pull/32154/files/16f380bd348c72695508a82d8501c9ee82a49959#diff-dc09df9d5422f6d9a1e4bdab0d4b0747bcd6bd5f5d4a561e6c07111e50afcc31R87 already checks that

File: projects/packages/videopress/src/class-access-control.php
Comment (millerf) [reply projects/packages/videopress/src/class-access-control.php:101]: It looks like they are checking if premium content block is loaded?

Comment (pgk) [reply projects/packages/videopress/src/class-access-control.php:101]: @jeherve @millerf yes, to my understanding there are situations it would not exist. Since all gating/subscription classes are part of only the Jetpack plugin, any other flavor (specifically Jetpack VideoPress plugin) would not contain those.

File: projects/packages/videopress/src/class-access-control.php
Comment (pgk) [reply projects/packages/videopress/src/class-access-control.php]: Works with read_post! Thanks

File: projects/packages/videopress/src/class-access-control.php
Comment (pgk) [reply projects/packages/videopress/src/class-access-control.php:86]: My reasons for checks I think I covered in the other https://github.com/Automattic/jetpack/pull/32154/files/16f380bd348c72695508a82d8501c9ee82a49959#r1354597036. 

> Should it also take module availability into account, and check if the Subscriptions module is active for example?

Maybe, let me check and report back

File: projects/packages/videopress/src/class-access-control.php
Comment (pgk) [reply projects/packages/videopress/src/class-access-control.php:107]: @millerf I do a require_once in the meantime for the file containing that function, so that check might return true.

File: projects/packages/videopress/src/class-access-control.php
Comment (pgk) [reply projects/packages/videopress/src/class-access-control.php]: after some thought, we are going to address this separately.

**Comment by pgk** [comment]
@roundhill @jeherve @millerf ready for a final rereview

File: projects/packages/videopress/src/class-access-control.php
Comment (jeherve) [inline projects/packages/videopress/src/class-access-control.php]: Nit-pick, but you can actually remove this phpcs exception at this point, your file name is valid. 👍

```suggestion
<?php
```

**Review by jeherve — COMMENTED** [review]
This tests well for me on WordPress.com Simple, and looks good. I'll let others chime in with their tests. 👍

File: projects/packages/videopress/src/client/lib/token-bridge/index.ts
Comment (millerf) [inline projects/packages/videopress/src/client/lib/token-bridge/index.ts:53]: TIL PHP has Promises ❤️

Comment (millerf) [inline projects/packages/videopress/src/client/lib/token-bridge/index.ts]: I think it would make things way clearer if ready() was returning a premise instead of a function. You're mixing callbacks and promises and it never gets good. What about this approach:

```

       
async function getSubscriberPlanIdIfExists( guid: VideoGUID ) : Promise<number> {  
     
     // Can also be defined outside
     const ready = new Promise( ( resolve ) => document.readyState != "loading" ? 
       resolve() : 
       document.addEventListener("DOMContentLoaded", resolve ) );
    
     await ready();
     if ( ! window.__guidsToPlanIds ) {
       return 0;
     }
     return window.__guidsToPlanIds[ guid ] || 0;
}

```
(syntax might be off as I wrote everything here directly)


Writing this, the last if actually useless:
```
   
async function getSubscriberPlanIdIfExists( guid: VideoGUID ) : Promise<number> {  
     
     // Can also be defined outside
     const ready = new Promise( ( resolve ) => document.readyState != "loading" ? 
       resolve() : 
       document.addEventListener("DOMContentLoaded", resolve ) );
    
     await ready();
     return window.__guidsToPlanIds[ guid ] || 0;
}
```

or, in short

```
const ready = new Promise( ( resolve ) => document.readyState != "loading" ? 
       resolve() : 
       document.addEventListener("DOMContentLoaded", resolve ) );
       
const getSubscriberPlanIdIfExists = async guid: VideoGUID => await ready().map( () =>  window.__guidsToPlanIds[ guid ] || 0 );      
```
(syntax might be off as I wrote everything here directly)

File: projects/packages/videopress/src/class-access-control.php
Comment (millerf) [inline projects/packages/videopress/src/class-access-control.php:150]: This is already defined 2 lines above...

File: projects/packages/videopress/src/class-access-control.php
Comment (millerf) [inline projects/packages/videopress/src/class-access-control.php]: unnecessary `break;`

File: projects/packages/videopress/src/class-access-control.php
Comment (millerf) [inline projects/packages/videopress/src/class-access-control.php]: Would it make sense to use
```
 $this->default_video_restriction_details( false );
 ```
 ?

Comment (millerf) [reply projects/packages/videopress/src/class-access-control.php]: You will get translations for free.

File: projects/packages/videopress/src/class-access-control.php
Comment (millerf) [inline projects/packages/videopress/src/class-access-control.php]: I think there is a helper function used now : `\Jetpack_Options::get_option( 'id' );` (it returns the cache site id)

File: projects/packages/videopress/src/class-access-control.php
Comment (millerf) [inline projects/packages/videopress/src/class-access-control.php]: As this is called every time a user restrictions is checked, shouldn't it be linked to the user and use
```
wpcom_json_api_request_as_user
```

File: projects/packages/videopress/src/class-access-control.php
Comment (millerf) [inline projects/packages/videopress/src/class-access-control.php]: I am struggling to understand why video  restriction details get updated on WPCOM every time `is_current_user_authed_for_video` is called...
At least shouldn't there be some cache (at least for the current request) to prevent multiple calls?

File: projects/packages/videopress/src/class-initializer.php
Comment (millerf) [inline projects/packages/videopress/src/class-initializer.php:317]: That can simplified as well

```
$script_content = "window.__guidsToPlanIds = { '$escaped_guid': $premium_block_plan_id, ...window.__guidsToPlanIds };"

```

See this screenshot (totally legit)

Screenshot 2023-10-27 at 12 44 50

**Review by millerf — COMMENTED** [review]
I left a few comments to maybe improve some JS.
Overall it looks good. 

I am only concerned about 
```
$this->update_video_restriction_details_on_wpcom( $guid, $restriction_details );
```
calling WPCOM everytime `is_current_user_authed_for_video` is called... But there might be some specifics to your logic that I might be missing. Might be worth some static cache for the time of the request?

Otherwise there is no issue with the Memberships interaction.

File: projects/packages/videopress/src/class-access-control.php
Comment (pgk) [reply projects/packages/videopress/src/class-access-control.php]: @millerf I thought about doing it on save but decided against it. Since the privacy setting is impacted by parent blocks, filters etc, it seems like the most "late-bound" place to do it. 

I will go ahead and remove the call entirely for this version of the feature. The impact if removed will be that subscriber-only videos will display a generic "video is private" error when not viewed within the site.

File: projects/packages/videopress/src/class-initializer.php
Comment (pgk) [reply projects/packages/videopress/src/class-initializer.php:317]: I was trying to be conservative with es6 features on this, since I am not sure where this will run.

File: projects/packages/videopress/src/class-access-control.php
Comment (pgk) [reply projects/packages/videopress/src/class-access-control.php]: removing this code entirely so things are more simple for now.

File: projects/packages/videopress/src/client/lib/token-bridge/index.ts
Comment (pgk) [reply projects/packages/videopress/src/client/lib/token-bridge/index.ts]: nice idea! Refactoring

**Comment by millerf** [comment]
Looks good to me once rebased 👍

**Comment by thedebian** [comment]
@millerf Can you approve the PR if it's ok now ?

**Review by millerf — APPROVED** [review]

**Pull request merged**
